api/odilo.py

```diff
@@ -0,0 +1,532 @@
+                    # Clearly our info is out of date.
+                    # self.update_licensepool(identifier.identifier)
+                    raise NoAvailableCopies()
+                elif error['id'] == 'PatronHasExceededCheckoutLimit':
+                    raise PatronLoanLimitReached()
+                else:
+                    raise CannotLoan(error['id'])
+        else:
+            return self.loan_info_from_odilo_checkout(licensepool.collection, response_json)
+
+    def loan_info_from_odilo_checkout(self, collection, checkout):
+        start_date = self.extract_date(checkout, 'startTime')
+        end_date = self.extract_date(checkout, 'endTime')
+
+        return LoanInfo(
+            collection,
+            DataSource.ODILO,
+            Identifier.ODILO_ID,
+            checkout['id'],
+            start_date,
+            end_date,
+            checkout['downloadUrl']
+        )
+
+    def checkin(self, patron, pin, licensepool):
+        checkout_id = licensepool.identifier.identifier
+        url = self.CHECKIN_ENDPOINT.replace("{checkoutId}", checkout_id) + "?patronId=" + patron
+
+        response = self.patron_request(patron, pin, url, method='POST')
+        if response.status_code == 404:
+            data = response.json()
+            if data and data['errors'] and len(response.json()['errors']) > 0:
+                error = data['errors'][0]
+                if error['id'] == 'CHECKOUT_NOT_FOUND':
+                    raise NotCheckedOut(error['description'])
+                else:
+                    raise CannotReturn(error['id'] + ': ' + error['description'])
+        else:
+            return response
+
+    @classmethod
+    def extract_date(cls, data, field_name):
+        if field_name not in data:
+            d = None
+        else:
+            # OdiloAPI dates are timestamps in milliseconds
+            d = datetime.datetime.fromtimestamp(float(data[field_name]) / 1000.0)
+        return d
+
+    def raise_exception_on_error(self, data, custom_error_to_exception={}):
+        if not data or 'errors' not in data or len(data['errors']) <= 0:
+            return
+
+        error = data['errors'][0]
+        error_code = error['id']
+        message = ('description' in error and error['description']) or ''
+        for d in custom_error_to_exception, self.error_to_exception:
+            if error in d:
+                raise d[error_code](message)
+
+    def get_loan(self, patron, pin, checkout_id):
+        url = self.CHECKOUT_GET.replace("{checkoutId}", checkout_id) + '?patronId=' + patron
+        response = self.patron_request(patron, pin, url)
+        if response.status_code == 200 and response.content:
+            data = response.json()
+            self.raise_exception_on_error(data)
+            return data
+        else:
+            return None
+
+    def get_hold(self, patron, pin, hold_id):
+        url = self.HOLD_GET.replace("{holdId}", hold_id)
```

Comment:
I'm not sure why you don't stick `patronId` on to this URL like you do with others.

Reply:
It was a mistake that has now been corrected, thank you !!